lstore/db.py

```diff
@@ -1,17 +1,131 @@
+                tables_metadata = json.load(file)
+
+                # loops through tables and adds them to the self.tables dictionary
+                for table_name, table_info in tables_metadata.items():
+                    table = Table(table_name, table_info["num_columns"], table_info["key_index"], self.path)
+                    self.tables[table_name] = table
+
+                    # retores table rid_index
+                    table.rid_index = table_info.get("rid_index", 0)
+
+                    # restores page directory for table - converts keys back into correct data type (string --> integer)
+                    table.page_directory = {}
+                    for k, v in table_info["page_directory"].items():
+                        table.page_directory[int(k)] = v
+
+                    # initialize indexes for table
+                    table.index = Index(table)
+                    
+                    # NO NEED TO MANUALLY DESERIALIZE FOR INDEXES
+                    # ---------------------------------------------
+                    # for column_idx, indexed_values in table_info["indexes"].items():
+                    #     column_idx = int(column_idx)
+
+                    #     # only restore if there were stored values
+                    #     if indexed_values:  
+                    #         restored_tree = OOBTree()
+                    #         for value, rids in indexed_values.items():
+
+                    #             # ensure column key and values are integers (JSON converts them to strings)
+                    #             restored_tree[int(value)] = {int(rid): True for rid in rids} 
+                    #         table.index.indices[column_idx] = restored_tree
+                    #     else:
+                    #         table.index.indices[column_idx] = None  # explicitly set to None for unindexed column
 
     def close(self):
-        pass
+        """Flushes dirty pages back to disk, saves table metadata and shuts down"""
+        if not self.path:
+            raise ValueError("Database path is not set. Use open() before closing.")
+        
+        tables_metadata = {}
+
+        for table_name, table in self.tables.items():
+            # flush dirty pages from table's bufferpool
+            for frame in table.bufferpool.frames:
+                if frame.dirty:
+                    frame.unload_page()
+
+            # serialize table metadata
+            tables_metadata[table_name] = table.serialize()
+
+            # clear each table's bufferpool to free memory
+            table.bufferpool = None
+
+        # save all tables' metadata to disk
+        tables_metadata_path = os.path.join(self.path, "tables.json")
+        with open(tables_metadata_path, "w", encoding="utf-8") as file:
+            json.dump(tables_metadata, file, indent=4)
+
+        # clear memory references
+        self.tables = {}
+
+    # def _serialize(self):
```

Comment:
No need to keep these comments since you're already doing most of the commented part in other parts.
If you really do need them you can make a separate branch